torch/csrc/autograd/functions/special.cpp

```diff
@@ -237,9 +255,28 @@ bool Eval::replaceSubgraph(const variable_list& inputs, const variable_list& _ou
   }
 
   // Rebase outputs.
+  auto this_shared = shared_from_this();
+  std::unordered_set<Variable*> repeated_outputs;
   for (auto & output : relevant_outputs) {
-    output.get()->_grad_fn = shared_from_this();
-    output.get()->output_nr = num_inputs++;
+    // This output is already rebased. This happens when there
+    // the same Variable has been returned multiple times, and
+    // is repeated in this list.
+    if (output.get()->_grad_fn.get() == this) {
```

Comment:
The combination of the the three states a Variable can be in {unrepeated, first-time repeated, many repeated} and the mutation of grad_fn/output_nr/next_functions as raw data structures makes this code hard to understand. Changes are probably not needed, but it does take a long time to figure out what is happening.